Lib/test/test_pyexpat.py

```diff
@@ -821,5 +824,181 @@ def start_element(name, _):
         self.assertEqual(started, ['doc'])
 
 
+class AttackProtectionTestBase(abc.ABC):
+    """
+    Base class for testing protections against XML payloads with
+    disproportionate amplification.
+
+    The protections being tested should detect and prevent attacks
+    that leverage disproportionate amplification from small inputs.
+    """
+
+    @staticmethod
+    def exponential_expansion_payload(ncols, nrows, text='.'):
+        """Create a billion laughs attack payload.
+
+        Be careful: the number of total items is pow(n, k), thereby
+        requiring at least pow(ncols, nrows) * sizeof(text) memory!
+        """
+        body = '\n'.join(
+            f'<!ENTITY row{i + 1} "{f"&row{i};" * ncols}">'
+            for i in range(nrows)
+        )
+        return f"""\
+<?xml version="1.0"?>
+<!DOCTYPE doc [
+<!ENTITY row0 "{text}">
+<!ELEMENT doc (#PCDATA)>
+{body}
+]>
+<doc>&row{nrows};</doc>"""
```

Comment:
`textwrap.dedent` could help readability here, e.g.:

```suggestion
        return dedent(f"""\
            <?xml version="1.0"?>
            <!DOCTYPE doc [
            <!ENTITY row0 "{text}">
            <!ELEMENT doc (#PCDATA)>
            {body}
            ]>
            <doc>&row{nrows};</doc>""")
```

Reply:
I don't really want an additional import. If I were to be honest, I would have inlined everything :')

Reply:
Ok, actually it's probably better to have a dedent but not that way. Let me do it locally.

Reply:
> I don't really want an additional import.

@picnixz what is the concern about `from textwrap import dedent` from a test file?

> Let me do it locally.

I'm not sure what that would look like, so curious to see it in action.